mungegithub/mungers/e2e/e2e.go

```diff
@@ -70,21 +77,142 @@ func (e *E2ETester) Stable() bool {
 			allStable = false
 		}
 	}
 	return allStable
 }
+
+const (
+	successString     = "SUCCESS"
+	expectedXMLHeader = "<?xml version=\"1.0\" encoding=\"UTF-8\"?>"
+)
+
+type finishedFile struct {
+	Result    string `json:"result"`
+	Timestamp uint64 `json:"timestamp"`
+}
+
+func checkFinishedStatus(job string, lastBuildNumber int) bool {
+	response, err := utils.GetFileFromJenkinsGoogleBucket(job, lastBuildNumber, "finished.json")
+	if err != nil {
+		glog.Errorf("Error while getting data for %v/%v/%v: %v", job, lastBuildNumber, "finished.json", err)
+		return false
+	}
+
+	defer response.Body.Close()
+	if response.StatusCode != 200 {
+		glog.Infof("Got a non-success response %v while reading data for %v/%v/%v", response.StatusCode, job, lastBuildNumber, "finished.json")
+		return false
+	}
+	result := finishedFile{}
+	body, err := ioutil.ReadAll(response.Body)
+	if err != nil {
+		glog.Errorf("Failed to read the response for %v/%v/%v: %v", job, lastBuildNumber, "finished.json", err)
+		return false
+	}
+	err = json.Unmarshal(body, &result)
+	if err != nil {
+		glog.Errorf("Failed to unmarshal %v: %v", string(body), err)
+	}
+	glog.Infof("%v", result)
+	return result.Result == successString
+}
+
+// GCSBasedStable is a version of Stable function that depends on files stored in GCS instead of Jenkis
+func (e *E2ETester) GCSBasedStable() bool {
+	for _, job := range e.JobNames {
+		lastBuildNumber, err := utils.GetLastestBuildNumberFromJenkinsGoogleBucket(job)
+		glog.Infof("Checking status of %v, %v", job, lastBuildNumber)
+		if err != nil {
+			glog.Errorf("Error while getting data for %v: %v", job, err)
+			continue
+		}
+		if !checkFinishedStatus(job, lastBuildNumber) {
+			return false
+		}
+	}
+
+	return true
+}
+
+// GCSWeakStable is a version of GCSBasedStable with a slightly relaxed condition.
+// This function says that e2e's are unstable only if there were real test failures
+// (i.e. there was a test that failed, so no timeouts/cluster startup failures counts),
+// or test failed for any reason 3 times in a row.
+func (e *E2ETester) GCSWeakStable() bool {
+	for _, job := range e.WeakStableJobNames {
+		lastBuildNumber, err := utils.GetLastestBuildNumberFromJenkinsGoogleBucket(job)
+		glog.Infof("Checking status of %v, %v", job, lastBuildNumber)
+		if err != nil {
+			glog.Errorf("Error while getting data for %v: %v", job, err)
+			continue
+		}
+		if checkFinishedStatus(job, lastBuildNumber) {
+			continue
+		}
+
+		// If we're here it means that build failed, so we need to look for a reason
+		// by iterating over junit_XX.xml files and look for failures
+		i := 0
+		for {
+			i++
+			path := fmt.Sprintf("artifacts/junit_%02d.xml", i)
+			response, err := utils.GetFileFromJenkinsGoogleBucket(job, lastBuildNumber, path)
+			if err != nil {
+				glog.Errorf("Error while getting data for %v/%v/%v: %v", job, lastBuildNumber, path, err)
+				continue
+			}
+			if response.StatusCode != 200 {
+				break
+			}
+			{
```

Comment:
Yeah... right. IIUC go have an early binding, so it should be OK with plain `defer` here... I hope.